### PR TITLE
Integrate MAML2Yaml

### PR DESCRIPTION
**Pull request by mingwli**
[AB#279561](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/279561)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6374)

### CODE REVIEW

File: src/docfx/config/ops/OpsPreProcessor.cs [hunk 1]
```diff
@@ -35,7 +42,7 @@ private void PreProcessMonoDocXml()
 
             using (Progress.Start("Preprocessing monodoc XML files"))
             {
-                lock (s_lock)
+                lock (s_ecma)
```
Comment (live1206) [inline src/docfx/config/ops/OpsPreProcessor.cs:45]: reuse `s_lock`?
I guess the lock is to assure that the repo content is locked for writing.
And `ecma` and `maml` are executed successively.

File: src/docfx/config/ops/OpsPreProcessor.cs [hunk 2]
```diff
@@ -60,14 +67,39 @@ private void PreProcessMonoDocXml()
...
+            if (_config.MAML2YamlMonikerMappingPath is null)
+            {
+                return;
+            }
+
+            using (Progress.Start("Preprocessing MAML markdown files"))
+            {
+                lock (s_maml)
+                {
+                    for (var index = 0; index < _config.MAML2YamlMonikerMappingPath.Length; index++)
```
Comment (live1206) [inline src/docfx/config/ops/OpsPreProcessor.cs:88]: why not just `foreach` on `_config.MAML2YamlMonikerMappingPath`? same in line 47

File: src/docfx/lib/markdown/IncludeExtension.cs [hunk 3]
```diff
@@ -97,7 +98,7 @@ private static void RemoveIncludeRenderer(IMarkdownRenderer renderer)
 
             using (InclusionContext.PushInclusion(file))
             {
-                var child = Markdown.Parse(content, pipeline);
+                var child = MarkdigMarkdown.Parse(content, pipeline);
```
[the remark this answers is not in the document]
Comment (mingwli) [reply src/docfx/lib/markdown/IncludeExtension.cs:101]: namespace/class name collision

File: src/docfx/config/ops/OpsPreProcessor.cs (hunk 2 above)
Comment (mingwli) [reply src/docfx/config/ops/OpsPreProcessor.cs:88]: need a `index` number for `sourcemap-{index}`

File: src/docfx/config/ops/OpsPreProcessor.cs (hunk 1 above)
Comment (mingwli) [reply src/docfx/config/ops/OpsPreProcessor.cs:45]: I thought one is treating `xml` files, one is treating `maml` files, there should be no intersection and can be run in parallel

File: src/docfx/config/ops/OpsConfigLoader.cs [hunk 4]
```diff
@@ -129,6 +143,20 @@ private static (JObject obj, string path, string name)[] GetDependencies(OpsConf
             return result.Count == 0 ? null : result;
         }
 
+        private static JArray? GetMAML2YamlMonikerPath(OpsDocsetConfig? docsetConfig, OpsConfig opsConfig, string buildSourceFolder)
```
[the remark this answers is not in the document]
Comment (mingwli) [reply src/docfx/config/ops/OpsConfigLoader.cs:146]: :/ I'm just following the `ECMA2Yaml` way

File: src/docfx/config/ops/OpsDocsetConfig.cs [hunk 5]
```diff
@@ -28,5 +28,9 @@ internal class OpsDocsetConfig
         [JsonProperty(nameof(ECMA2Yaml))]
         [JsonConverter(typeof(OneOrManyConverter))]
         public ECMA2YamlRepoConfig[]? ECMA2Yaml { get; private set; }
+
+        [JsonProperty("monikerPath")]
+        [JsonConverter(typeof(OneOrManyConverter))]
+        public string[]? MAML2YamlMonikerMappingPath { get; set; }
```
[the remark this answers is not in the document]
Comment (mingwli) [reply src/docfx/config/ops/OpsDocsetConfig.cs:34]: > Keep C# name the same as JSON name

`monikerPath` feels kinda too general to me, it's hard to link it with `MAML2Yaml`.
btw, what's the cons of using different name?

File: src/docfx/config/ops/OpsConfigLoader.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/config/ops/OpsConfigLoader.cs:140]: Change `ECMA2Yaml` as well?

File: src/docfx/lib/json/JsonUtility.cs [hunk 6]
```diff
@@ -297,6 +297,14 @@ public static JToken DeepClone(JToken? token)
             throw new NotSupportedException();
         }
 
+        public static void AddRange(this JArray container, JArray arr)
```
Comment (yufeih) [inline src/docfx/lib/json/JsonUtility.cs:300]:
```suggestion
        public static void AddRange(this JArray container, IEnumerable<JToken> arr)
```

File: src/docfx/config/ops/OpsConfig.cs [hunk 7]
```diff
@@ -29,5 +29,9 @@ internal class OpsConfig
         [JsonProperty(nameof(ECMA2Yaml))]
         [JsonConverter(typeof(OneOrManyConverter))]
         public ECMA2YamlRepoConfig[]? ECMA2Yaml { get; private set; }
+
+        [JsonProperty("monikerPath")]
+        [JsonConverter(typeof(OneOrManyConverter))]
+        public string[]? MonikerPath { get; set; }
```
Comment (herohua) [inline src/docfx/config/ops/OpsConfig.cs:35]: >MonikerPath [](start = 25, length = 11)

could other reference language also uses this configuration? the name seems too vague to be associated with PowerShell

Comment (mingwli) [reply src/docfx/config/ops/OpsConfig.cs:35]: If there's a name collision, we may:
- either ask plugin team to come up with a new name
- consume the plugin name in `customized_tasks` to distinguish, which I wouldn't propose :/
Simply change the `Property` name won't change anything

**Pull request merged**